libknet/transport_udp.c

```diff
@@ -0,0 +1,177 @@
+#include "config.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <pthread.h>
+#include <math.h>
+#include <sys/epoll.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h>
+#include <malloc.h>
+#include <qb/qblist.h>
+
+#include "libknet.h"
+#include "host.h"
+#include "link.h"
+#include "logging.h"
+#include "common.h"
+#include "transports.h"
+
+#define KNET_PMTUD_UDP_OVERHEAD 8
+
+typedef struct udp_handle_info {
+	knet_handle_t knet_h;
+	struct qb_list_head links_list;
+} udp_handle_info_t;
+
+typedef struct udp_link_info {
+	knet_transport_link_t transport;
+	knet_handle_t knet_handle;
+	int socket_fd;
+	udp_handle_info_t *handle_info;
+	struct qb_list_head list;
+	struct sockaddr_storage local_address;
+} udp_link_info_t;
+
+static int udp_handle_allocate(knet_handle_t knet_h, knet_transport_t *transport)
+{
+	udp_handle_info_t * handle_info;
+
+	handle_info = malloc(sizeof(udp_handle_info_t));
+	if (!handle_info) {
+		return -1;
+	}
+	handle_info->knet_h = knet_h;
+	qb_list_init(&handle_info->links_list);
+
```

Comment:
couple of bits. If transports need to malloc, then please malloc / init all of them at knet_handle_new (perhaps you already do, but i just want to make sure ;))

The general idea is that the only malloc at runtime is to add new hosts. Everything underneath should already be all initialized. Idea of either it works or it doesn´t and there is only one point where it can fail allocating memory.

Also, qb_list should go :-)